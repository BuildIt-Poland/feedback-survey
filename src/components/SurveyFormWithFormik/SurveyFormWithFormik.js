import { withFormik } from 'formik';

import { REQUIRED_QUESTION } from '../../constatnts/messages';
import { saveSurvey } from '../../services/api';
import SurveyForm from '../SurveyForm';

const questionValidate = (values, props) => {
  const { questions } = props.surveyData;
  let errors = {};

  questions
    .filter(item => item.required)
    .filter(item => !values[item.name])
    .forEach(item => {
      errors[item.name] = REQUIRED_QUESTION;
    });

  return errors;
};

const setInitialValues = props => {
  const { questions } = props.surveyData;
  const init = {};

  questions.map(item => (init[item.name] = ''));

  return init;
};

const removeEmptyValues = obj => {
  Object.keys(obj).forEach(key => obj[key] === '' && delete obj[key]);
};

const handleSubmit = (values, { props, setStatus }) => {
  const { surveyId, employeeName } = props;
  const surveyStatus = {
    isSuccessSave: false,
    saveErrorMessage: '',
    isSubmitted: false
  };
  removeEmptyValues(values);
  const data = {
    surveyId,
    employeeName,
    answers: values
  };

  const saveData = async data => {
    try {
      await saveSurvey(data);
      surveyStatus.isSuccessSave = true;
    } catch (error) {
      surveyStatus.saveErrorMessage = error.message;
    }
    surveyStatus.isSubmitted = true;
    setStatus(surveyStatus);
  };

  saveData(data);
};

const SurveyFormWithFormik = withFormik({
  mapPropsToValues: setInitialValues,
  validate: questionValidate,
  handleSubmit: handleSubmit
})(SurveyForm);

export default SurveyFormWithFormik;
