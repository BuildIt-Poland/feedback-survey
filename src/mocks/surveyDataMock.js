export const surveyDataMock = {
  surveyData: {
    questions: [
      {
        id: '4',
        type: '',
        content: 'Test question?',
        required: false,
        name: 'testName',
        answerType: 'kindOfFeedback'
      }
    ],
    answerTypes: [
      {
        type: 'kindOfFeedback',
        values: ['monthly', 'quaterly', 'annual']
      }
    ]
  },
  errors: {},
  values: {}
};
