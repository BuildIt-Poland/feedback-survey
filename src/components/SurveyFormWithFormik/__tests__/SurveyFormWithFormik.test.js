import React from 'react';
import { create } from 'react-test-renderer';

import SurveyFormWithFormik from '../SurveyFormWithFormik';
import { surveyDataMock } from '../../../mocks/surveyDataMock';

describe('COMPONENT - SurveyFormWithFormik', () => {
  it('renders correctly', () => {
    const component = create(<SurveyFormWithFormik {...surveyDataMock} />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
