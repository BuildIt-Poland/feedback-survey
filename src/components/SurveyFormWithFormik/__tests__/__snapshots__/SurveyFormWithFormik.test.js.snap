// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`COMPONENT - SurveyFormWithFormik renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="StyledGrommet-sc-19lkkz7-0 eUfLDp"
  >
    <div
      className="sc-bZQynM eCWoqb"
    >
      <div
        className="sc-EHOje dYFYCA FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy"
      >
        <label
          className="StyledText-sc-1sadyjn-0 bMXxti"
        >
          Test question?
        </label>
        <div
          className="StyledBox-sc-13pk1d4-0 gZSidW"
        >
          <div
            className="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 yIdXy"
          >
            <input
              autoComplete="off"
              className="StyledTextInput-sc-1x30a0s-0 cyWwXE"
              name="testName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              value=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="StyledBox-sc-13pk1d4-0 dNgQiZ"
  >
    <button
      className="sc-bdVaJa dGAbuL StyledButton-sc-323bzc-0 bmExQD"
      onBlur={[Function]}
      onFocus={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      size="large"
      type="submit"
    >
      submit
    </button>
  </div>
</form>
`;
